test/unit/app/controllers/permissions/permissions-controller-test.js

```diff
@@ -0,0 +1,1392 @@
+        'should not throw on non-existing request'
+      )
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals still empty after request',
+      )
+    })
+
+    it('rejects single existing request', async function () {
+
+      assert.rejects(
+        mockRequestUserApproval(REQUEST_IDS.a),
+        ERRORS.rejectPermissionsRequest.rejection(),
+        'rejects as expected'
+      )
+
+      await permController.rejectPermissionsRequest(REQUEST_IDS.a)
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals empty after rejection',
+      )
+    })
+
+    it('rejects requests regardless of order', async function () {
+
+      assert.rejects(
+        mockRequestUserApproval(REQUEST_IDS.b),
+        ERRORS.rejectPermissionsRequest.rejection(),
+        'rejects as expected'
+      )
+
+      assert.rejects(
+        mockRequestUserApproval(REQUEST_IDS.c),
+        ERRORS.rejectPermissionsRequest.rejection(),
+        'rejects as expected'
+      )
+
+      // reject out of order
+      await permController.rejectPermissionsRequest(REQUEST_IDS.c)
+      // add a non-existing request to the mix
+      await permController.rejectPermissionsRequest(REQUEST_IDS.a)
+      await permController.rejectPermissionsRequest(REQUEST_IDS.b)
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals empty after approval',
+      )
+    })
+  })
+
+  describe('_requestPermissions', function () {
+
+    let permController
+
+    beforeEach(function () {
+      permController = initPermController()
+    })
+
+    it('requests the given permissions and grants them on user approval', async function () {
+
+      permController._requestPermissions(
+        ORIGINS.a, PERMS.requests.eth_accounts()
+      )
+        .then(async (result) => {
+
+          // this is the last thing that will happen in this test
+
+          assert.ok(
+            result.length === 1 && result[0].parentCapability === 'eth_accounts',
+            'single eth_accounts permission should have been granted'
+          )
+
+          const accounts = await permController.getAccounts(ORIGINS.a)
+          assert.deepEqual(
+            accounts, ACCOUNT_ARRAYS.a, 'origin should have correct accounts'
+          )
+        })
+        .catch(() => {
+          assert.fail('promise should not reject')
+        })
+
+      const accounts = await permController.getAccounts(ORIGINS.a)
+
+      assert.deepEqual(accounts, [], 'origin should not have any accounts')
+
+      assert.equal(
+        permController.pendingApprovals.size, 1,
+        'perm controller should have single pending approval',
+      )
+
+      const id = permController.pendingApprovals.keys().next().value
+      const request = PERMS.approvedRequest(id, PERMS.requests.eth_accounts())
+
+      permController.approvePermissionsRequest(request, ACCOUNT_ARRAYS.a)
```

Comment:
Unfortunately the options for wholly stubbing-out `rpc-cap` are limited - you'd have to use `proxyquire` or refactor the controller to pass that in as an option as well. But if we're OK with it being constructed then replaced, we can do this:

```
permController.permissions = { providerMiddlewareFunction: sinon.spy() }
```

Then the test would be a s simple as:

```
const pendingRequest = permController._requestPermissions(ORIGINS.a, PERMS.requests.eth_accounts())

const _end = permController.permissions.providerMiddlewareFunction.getCall(0).args[4]
_end()

const result = await pendingRequest
// assert things on result
```

I suspect we could treat these other tests the same way, to varying degrees of difficulty. But this seemed like an especially good candidate.

Reply:
I struggle with this. On one hand, your suggestion vastly improves test encapsulation, and doesn't rely on implementation details of `rpc-cap`. On the other hand, these methods rely on implementation details of `rpc-cap`, and we want to know if a change in the latter breaks the former!

Your specific suggestion doesn't work, because the actual `rpc-cap` middleware is what produces the the `result`, so there's to assert on it. If I'm writing a test that only tells me that a method on `rpc-cap` is being called... that's still valuable, but the value is IMO diminished relative to what we have.

I don't know, what do you think?

Related: https://github.com/MetaMask/metamask-extension/pull/7969#discussion_r391893844

Reply:
>On the other hand, these methods rely on implementation details of rpc-cap, and we want to know if a change in the latter breaks the former!

Do they though 🤔 This should be the public interface of `rpc-cap` right?

You're still right that we're relying on the behavior of our mock/stub being correct. It just feels silly to go through such lengths to test a very simple pass-this-in-then-return-it type method. I don't know. Maybe it does make sense to treat `rpc-cap` as internal rather than external because of how interwoven they are.

Reply:
Yeah, it's ambiguous, but it sounds to me like the solution will have to wait. Let's leave this one for now?

Reply:
Sure, I suppose so.